### PR TITLE
[4.0] Adapt core extensions to new MVC structure

### PR DESCRIPTION
**Pull request by laoneo**
This is the follow up pr from #17615 and #17614. It contains also the merge commits of this pr's, otherwise the 4 branch will be broken.

I will do another one when #17681 and #17624 are merged.

### CODE REVIEW

**Comment by laoneo** [comment]
Conflicts fixed...

**Comment by wilsonge** [comment]
OK I made a start at this with https://github.com/joomla/joomla-cms/commit/23ac2a53cd2db219b95967717b5c4c00e0ddae94 when I did my merge cause I didn't see this at the time - but haven't had time to keep working on it the last couple of days. I've pushed up what I had done - what's easier - for me to keep doing it or you to get this in sync or?

**Comment by laoneo** [comment]
This one is ready to get merged.

File: libraries/src/MVC/Controller/FormController.php [hunk 1]
```diff
@@ -504,14 +503,14 @@ protected function getRedirectToListAppend()
...
-	 * @param   array      $validData  The validated data.
+	 * @param   \JModelLegacy  $model      The data model object.
+	 * @param   array          $validData  The validated data.
 	 *
 	 * @return  void
 	 *
 	 * @since   1.6
 	 */
-	protected function postSaveHook(BaseModel $model, $validData = array())
+	protected function postSaveHook(\JModelLegacy $model, $validData = array())
```
Comment (wilsonge) [inline libraries/src/MVC/Controller/FormController.php]: Why are we reverting this to `\JModelLegacy`?

File: components/com_config/Model/Cms.php [hunk 2]
```diff
@@ -70,7 +72,7 @@ public function __construct($config = array())
 
 			if (!preg_match('/(.*)Model/i', get_class($this), $r))
 			{
-				throw new \Exception(\JText::_('JLIB_APPLICATION_ERROR_MODEL_GET_NAME'), 500);
+				throw new \Exception(JText::_('JLIB_APPLICATION_ERROR_MODEL_GET_NAME'), 500);
```
Comment (wilsonge) [inline components/com_config/Model/Cms.php]: This is wrong too?

File: administrator/components/com_fields/Model/Group.php [hunk 3]
```diff
@@ -223,17 +222,17 @@ protected function preprocessForm(\JForm $form, $data, $group = 'content')
 		if ($section !== null)
 		{
 			// Looking first in the component models/forms folder
-			$path = \JPath::clean(JPATH_ADMINISTRATOR . '/components/' . $component . '/models/forms/fieldgroup/' . $section . '.xml');
+			$path = JPath::clean(JPATH_ADMINISTRATOR . '/components/' . $component . '/models/forms/fieldgroup/' . $section . '.xml');
```
Comment (wilsonge) [inline administrator/components/com_fields/Model/Group.php]: These ones too :)

File: administrator/components/com_fields/Controller/Controller.php [hunk 4]
```diff
@@ -52,7 +49,8 @@ public function display($cachable = false, $urlparams = false)
 		if ($vName == 'field' && !$this->checkEditId('com_fields.edit.field', $id))
 		{
 			// Somehow the person just went to the form - we don't allow that.
-			$this->setMessage(\JText::sprintf('JLIB_APPLICATION_ERROR_UNHELD_ID', $id), 'error');
+			$this->setError(\JText::sprintf('JLIB_APPLICATION_ERROR_UNHELD_ID', $id));
```
Comment (wilsonge) [inline administrator/components/com_fields/Controller/Controller.php]: And this

File: administrator/components/com_content/Controller/Articles.php [hunk 5]
```diff
@@ -85,7 +84,7 @@ public function featured()
 		else
 		{
 			// Get the model.
-			/** @var \Joomla\Component\Content\Administrator\Model\Article $model */
+			/** @var ContentModelArticle $model */
```
Comment (wilsonge) [inline administrator/components/com_content/Controller/Articles.php]: This

**Comment by wilsonge** [comment]
Think that's it once you get those last 3 :) Nice stuff! Life saving!

**Pull request merged**